src/stan/lang/generator/write_begin_all_dims_col_maj_loop.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef STAN_LANG_GENERATOR_WRITE_BEGIN_ALL_DIMS_COL_MAJ_LOOP_HPP
+#define STAN_LANG_GENERATOR_WRITE_BEGIN_ALL_DIMS_COL_MAJ_LOOP_HPP
+
+#include <stan/lang/ast.hpp>
+#include <stan/lang/generator/constants.hpp>
+#include <stan/lang/generator/generate_expression.hpp>
+#include <stan/lang/generator/generate_indent.hpp>
+#include <ostream>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace lang {
+
+    /**
+     * Generate the openings of a sequence of zero or more for loops
+     * corresponding to all dimensions of a variable, with the
+     * specified indentation level writing to the specified stream.
+     * If specified, declare named size_t variable for each dimension
+     * which avoids re-evaluation of size expression on each iteration.
+     *
+     * Indexing order is column major, nesting is innermost to outermost
+     * e.g., 3-d array of matrices indexing order:  col, row, d3, d2, d1
+     *
+     * @param[in] var_decl variable declaration
+     * @param[in] declare_size_vars if true, generate size_t var decls
+     * @param[in] indent indentation level
+     * @param[in,out] o stream for generating
+     */
+    void write_begin_all_dims_col_maj_loop(const block_var_decl& var_decl,
+                                           bool declare_size_vars,
+                                           int indent, std::ostream& o) {
+      std::string name(var_decl.name());
+      expression arg1(var_decl.type().arg1());
+      expression arg2(var_decl.type().arg2());
+      if (var_decl.type().is_array_type()) {
+        arg1 = var_decl.type().array_contains().arg1();
+        arg2 = var_decl.type().array_contains().arg2();
+      }
+      std::vector<expression> ar_var_dims = var_decl.type().array_lens();
+
+      if (!is_nil(arg2)) {
```

Comment:
[optional]
This is another case where the array dimensions are being fiddled together with innner matrix/vector dimensions.

[optional]
I think it might be easier to instead just pull all the dimensions out and do a consistent loop over all of them rather than having all these conditionals here.  If so, this'd be another motivation for the complete dimension pulling method.  I just worry it won't be so useful with tuples, where tuples nest in arrays, nest in tuples and so forth.